File: environments_harbor/redis_pub_sub_message_loss_medium/environment/subscriber.py
#!/usr/bin/env python3

import redis
import time
import sys

def main():
    try:
        # Create Redis client connection
        print("Connecting to Redis server at localhost:6379...")
        r = redis.Redis(host='localhost', port=6379, decode_responses=True)
        
        # Test connection
        r.ping()
        print("Successfully connected to Redis!")
        
        # Simulate connection setup time
        time.sleep(0.5)
        
        # Create pubsub object and subscribe
        pubsub = r.pubsub()
        pubsub.subscribe('system_logs')
        
        print("Subscribed to 'system_logs' channel. Waiting for messages...")
        print("Press Ctrl+C to stop.")
        
        message_count = 0
        
        # Listen for messages
        for message in pubsub.listen():
            if message['type'] == 'message':
                message_count += 1
                print(f"Message {message_count} received: {message['data']}")
            elif message['type'] == 'subscribe':
                print(f"Subscription confirmed to channel: {message['channel']}")
                
    except redis.ConnectionError as e:
        print(f"ERROR: Could not connect to Redis server: {e}", file=sys.stderr)
        sys.exit(1)
    except KeyboardInterrupt:
        print(f"\nShutting down subscriber. Total messages received: {message_count}")
        sys.exit(0)
    except Exception as e:
        print(f"ERROR: Unexpected error occurred: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: environments_harbor/nats_streaming_queue_group_hard/environment/subscriber.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	stan "github.com/nats-io/stan.go"
)

// Buggy subscriber implementation - contains 5 bugs that cause duplicate message processing
func main() {
	// Bug 4: Wrong cluster ID - using incorrect cluster ID
	clusterID := "test-cluster"
	clientID := fmt.Sprintf("subscriber-%d", time.Now().Unix())
	
	// Connect to NATS Streaming server
	sc, err := stan.Connect(
		clusterID,
		clientID,
		stan.NatsURL("nats://localhost:4222"),
	)
	if err != nil {
		log.Fatalf("Failed to connect to NATS Streaming: %v", err)
	}
	defer sc.Close()

	log.Printf("Connected to NATS Streaming as client: %s", clientID)

	subject := "orders"
	
	// Message handler function
	msgHandler := func(msg *stan.Msg) {
		log.Printf("Received message: %s (sequence: %d)", string(msg.Data), msg.Sequence)
		
		// Process the message
		processOrder(msg.Data)
		
		// Bug 3: Missing message acknowledgment - msg.Ack() not called
		log.Printf("Processed message sequence: %d", msg.Sequence)
	}

	// Bug 1: Missing queue group name - empty string causes all instances to receive messages
	queueGroup := ""
	
	// Subscribe to the subject with buggy options
	// Bug 2: Wrong subscription option - SetManualAckMode(false) means auto-ack
	// Bug 5: StartWithLastReceived() is wrong for queue groups - causes redelivery issues
	sub, err := sc.QueueSubscribe(
		subject,
		queueGroup,
		msgHandler,
		stan.DurableName("order-processor"),
		stan.SetManualAckMode(false),
		stan.StartWithLastReceived(),
		stan.AckWait(30*time.Second),
	)
	if err != nil {
		log.Fatalf("Failed to subscribe: %v", err)
	}
	defer sub.Unsubscribe()

	log.Printf("Subscribed to subject '%s' with queue group '%s'", subject, queueGroup)
	log.Println("Waiting for messages... (Press Ctrl+C to exit)")

	// Setup signal handling to gracefully shutdown
	signalChan := make(chan os.Signal, 1)
	signal.Notify(signalChan, os.Interrupt, syscall.SIGTERM)
	
	// Block until signal received
	<-signalChan
	
	log.Println("Shutting down subscriber...")
}

// processOrder simulates order processing
func processOrder(data []byte) {
	// Simulate processing time
	time.Sleep(100 * time.Millisecond)
	
	// In production, this would:
	// - Parse the order data
	// - Validate the order
	// - Store in database
	// - Trigger fulfillment workflow
	log.Printf("Processing order data: %s", string(data))
}